sw/device/lib/dif/dif_rv_timer.c

```diff
@@ -0,0 +1,364 @@
+    return kDifRvTimerBadArg;
+  }
+
+  // We need to read a monotonically increasing, volatile uint64. To do so,
+  // we first read the upper half, then the lower half. Then, we check if the
+  // upper half reads the same value again. If it doesn't, it means that the
+  // lower half overflowed and we need to re-take the measurement.
+  while (true) {
+    uint32_t upper = mmio_region_read32(
+        timer->base_addr,
+        reg_for_hart(hart_id, RV_TIMER_TIMER_V_UPPER0_REG_OFFSET));
+    uint32_t lower = mmio_region_read32(
+        timer->base_addr,
+        reg_for_hart(hart_id, RV_TIMER_TIMER_V_LOWER0_REG_OFFSET));
+
+    uint32_t overflow_check = mmio_region_read32(
+        timer->base_addr,
+        reg_for_hart(hart_id, RV_TIMER_TIMER_V_UPPER0_REG_OFFSET));
+
+    if (upper == overflow_check) {
+      *out = (((uint64_t)upper) << 32) | lower;
+      return kDifRvTimerOk;
+    }
+  }
+}
+
+dif_rv_timer_result_t dif_rv_timer_arm(const dif_rv_timer_t *timer,
+                                       uint32_t hart_id, uint32_t comp_id,
+                                       uint64_t threshold) {
+  if (timer == NULL || hart_id >= timer->config.hart_count ||
+      comp_id >= timer->config.comparator_count) {
+    return kDifRvTimerBadArg;
+  }
+
+  uint32_t lower = threshold;
+  uint32_t upper = threshold >> 32;
+
+  ptrdiff_t lower_reg =
+      reg_for_hart(hart_id, RV_TIMER_COMPARE_LOWER0_0_REG_OFFSET) +
+      (sizeof(uint64_t) * comp_id);
+  ptrdiff_t upper_reg =
+      reg_for_hart(hart_id, RV_TIMER_COMPARE_UPPER0_0_REG_OFFSET) +
+      (sizeof(uint64_t) * comp_id);
+
+  // First, set the upper register to the largest value possible without setting
+  // off the alarm; this way, we can set the lower register without setting
+  // off the alarm.
+  mmio_region_write32(timer->base_addr, upper_reg, UINT32_MAX);
+
+  // This can't set off the alarm because of the value we set above.
+  mmio_region_write32(timer->base_addr, lower_reg, lower);
+  // Finish writing the new value; this may set off an alarm immediately.
+  mmio_region_write32(timer->base_addr, upper_reg, upper);
+
+  return kDifRvTimerOk;
+}
+
+// Computes the appropriate register for a particular interrupt register, given
+// a number of comparators.
+//
+// Currently, all IRQ registers are placed after the comparator registers,
+// so the register offsets given by HW are not useable. The offsets need to be
+// compensated for by adding a double word for each comparator after the first
+// one.
+//
+// We also do not handle the case when comparator_count > 32, which would cause
+// multiple interrupt registers to be generated.
+static ptrdiff_t irq_reg_for_hart(uint32_t hart_id, uint32_t comparators,
+                                  ptrdiff_t reg_offset) {
+  ptrdiff_t base = reg_for_hart(hart_id, reg_offset);
+  ptrdiff_t extra_comparator_offset = sizeof(uint64_t) * (comparators - 1);
+  return base + extra_comparator_offset;
+}
```

Comment:
Yes, unfortunately. The register offsets in the .h file are correct for earlgrey, but not in general: they assume one hart and one comparator per. To be fully general, we need to know the number of comparators, since the IRQ registers are placed after the comparators. I've spoken to @eunchan about switching the order (IRQs, then compartors), but I don't think we ever got a consensus there.

Reply:
I see, but for an instance with two comparators, I would expect the registers of both of the comparators to be before the irq registers and the values of `RV_TIMER_INTR_{ENABLE0,STATE0,TEST0}_REG_OFFSET` to be shifted accordingly.

Reply:
I talked about this with @mcy offline: It looks like the best solution would be to move irq registers before the comparator registers (since their number is configurable). Without this we don't think there is a great solution:
- Adding the offset for comparators here means that if this DIF library is compiled for an rv_timer with N comparators, it can be used for rv_timers with >= N comparators.
- Not adding the offset for comparators here means that if this DIF library is compiled for an rv_timer with N comparators, it can be used for rv_timers with 1..N comparators.
I will create an issue to see if we can move the irq registers before the comparator registers.

@mcy, if things stay the same I think it would be a good idea to add a check in init to avoid surprises.

Reply:
I realized that this can be written to be fully genericaly, so I've gone ahead and updated it as such. =)

Reply:
I think @alphan has filed this as a reggen issue? Would be good to link it here.